### DISCUSSION TITLE
Where to statically fetch site-wide data?

### DISCUSSION THREAD

**Discussion by tylermcrobert** [vercel/next.js#10909] [Help] [answered]
(9.3.0) I have data that my entire site relies upon (nav, footer) that I really only want to fetch at build time. Is there a way to do this within _app.js. From what I understand, `getStaticProps` doesn't work at the component level or within `_app`

**Comment by MickCoelho** [comment]
I'm running into the same issue. In the meantime, the only solution I found is to use `getInitialProps` (which apparently doesn't opt-out children pages from SSG).

**Reply by rafaelalmeidatk** [reply]
~It does: https://err.sh/next.js/opt-out-auto-static-optimization~

**Reply by timneutkens** [reply]
Using `getStaticProps` will force the page to be rendered statically, even if _app has getInitialProps, this is mostly done for backwards compat and incremental adoption paths to static rendering.

_app will support getStaticProps soon, see https://github.com/zeit/next.js/discussions/10949

**Reply by rafaelalmeidatk** [reply]
@timneutkens I am not sure I understood the behavior correctly. If I have `getInitialProps` on _app and a page using `getStaticProps`, how will _app render some content correctly if you don't have the props from `getInitialProps`?

I am playing with different pages setups and what I think is happening is: during build time, for SSG pages, `getInitialProps` from _app runs and the result is frozen, all SSG pages will use these props that were aquired during build time. Is this correct?

This is the behavior I am seeing on pages but not on the root page (`/`). It is the only one where `getInitialProps` from _app is still running in a weird way (the page is still static, looks like it is running in background in the server?), so after refreshing the page the props are still being updated without rebuilds.

**Reply by timneutkens** [reply]
> I am playing with different pages setups and what I think is happening is: during build time, for SSG pages, getInitialProps from _app runs and the result is frozen

Correct

> This is the behavior I am seeing on pages but not on the root page (/). It is the only one where getInitialProps from _app is still running in a weird way (the page is still static, looks like it is running in background in the server?), so after refreshing the page the props are still being updated without rebuilds.

This has been fixed in next@canary, will be on 9.3.1

**Reply by rafaelalmeidatk** [reply]
Thanks for the explanation, Tim 🙏

**Reply by techrome** [reply]
> This has been fixed in next@canary, will be on 9.3.1

I'm still running into this issue with Next 10. "/" root page still calls getInitialProps from _app.js every time I navigate to this page ( I mean the result is not frozen, even with getStaticProps and revalidate). All other pages work perfectly fine, but not the index page.

I also noticed that it does not necessarily call it every time, but with 1 second window (even though I have 20 seconds revalidate), so if I call it 2 times a row within 1 second, I will still get old result. Maybe this was done intentionally?

**Comment by charles4221** [comment]
@MickCoelho it definitely still opts all pages out of potential static optimisation :(
I was hoping I could use either `getStaticProps` or `getServerSideProps` in `_app.js` for the same use-case of fetching global data like Header/Footer/Navigation, but neither works. So it looks like if I want to use static optimisation, I'll have to add my API calls which bring in the global data to **every page** which is super annoying.

**Reply by MickCoelho** [reply]
You're right, it opts out static optimization (even if the graph at the end of `next export` says otherwise).
I feel you, I just spent hours trying to do the exact same thing. Also, if the logic of the _app.jsx depends on some global content (which is the case in my app), the logic has to be duplicated on all the pages.

**Reply by charles4221** [reply]
In my case, prior to v9.3 I was using `_app.js` to fetch that global data and then storing it to a React Context object like `<GlobalContext.Provider value={globalData}>` at the App level, meaning I would then be able to easily consume any of the global data from any level of the tree, regardless of the current page or component(s). Not really sure how I can replace this and take advantage of the new SSG features.

**Reply by MickCoelho** [reply]
@charles4221 how exactly can you tell it opts out of SSG? Is that using the graph at the end of the `next export`? 
![image](https://user-images.githubusercontent.com/2513247/76441460-ad21f600-637c-11ea-983f-5ae7b20f7105.png)

**Reply by charles4221** [reply]
@MickCoelho I was getting the usual warning at build time that says “you have opted out of all static optimisation because of getInitialProps on App” or something to that effect. From what @timneutkens said below though, I guess this warning is a false positive?

**Reply by Vadorequest** [reply]
I believe it's a false positive that comes from 9.2 where it was true back then and hasn't been changed.

**Reply by ledenis** [reply]
If we use this approach, the Static Optimization Indicator [1] seems not to show up on the page, even though the graph at the `next build` says it's SSG'd. Can anyone confirm this?

[1] This icon : 
![next-static-optimization-indicator](https://user-images.githubusercontent.com/2855723/85315552-8c108b80-b4bb-11ea-9165-d415a25176a6.png)

**Comment by leerob** [comment, marked as answer]
Stumbled on this old discussion, but this is now possible with the App Router.

https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating
